### DISCUSSION TITLE
Why is the binary compiled differently each time?

### DISCUSSION THREAD

**Discussion by Bach234** [intel/llvm#6391] [Q&A] [answered]
When I compile with:
`clang -fsycl -c test.cpp -o test-0.o`
`clang -fsycl -c test.cpp -o test-1.o`
test-0.o and test-1.o always aren't same, they have random number in *.o files like this:
test-0.o: `GLOBAL__sub_I_test_aef76b.cpp`
test-1.o: `GLOBAL__sub_I_test_2340a8.cpp`
This feature doesn't exist in `g++`, I don't know why?
![0](https://user-images.githubusercontent.com/50255983/177143470-c46757f1-b0c8-444a-8c08-105bc66d0e08.PNG)
![1](https://user-images.githubusercontent.com/50255983/177143481-78df84a7-53e2-455d-acbe-f8a4d14dab4a.PNG)

**Comment by al42and** [comment, marked as answer]
I'm not an expert, but, in short, when compiling SYCL code, clang has to do multiple stages and generate several intermediate files with such randomized names.

You can read more here: https://github.com/intel/llvm/blob/sycl/sycl/doc/design/CompilerAndRuntimeDesign.md

Or you can add `-v` switch to clang and see it for yourself.

Note: `-save-temps` flag disables this randomization and allows reproducible builds.

**Reply by Bach234** [reply]
Thank you for your help! Your answer is very useful. I will try it soon.

**Reply by AlexeySachkov** [reply]
To further add here: the "faulty" step is so-called integration footer. You can find more details in my reply [here](https://github.com/intel/llvm/issues/7385#issuecomment-1316749127), but to summarize: our compiler replaces your main file with an intermediate (to append some code at the end of your `.cpp` file) and that makes the name unique for each compiler invocation. This behavior can be disabled with `-fno-sycl-use-footer`, but with that flag you can't use some of SYCL core and extension features.

**Reply by Bach234** [reply]
Thanks! This difference is from `SourceFileName` in `llvm/lib/IR/Module.cpp`, we can delete `SourceFileName` to solve this problem.
